// ng-spy/esm2020/lib/scroll-spy.module.mjs
import { NgModule } from '@angular/core';
import { SpyTargetDirective } from './spy-target.directive';
import { SpyOnDirective } from './spy-on.directive';
import * as i0 from "@angular/core";
export class ScrollSpyModule {
}
ScrollSpyModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: ScrollSpyModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ScrollSpyModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.1.0", ngImport: i0, type: ScrollSpyModule, declarations: [SpyTargetDirective, SpyOnDirective], exports: [SpyTargetDirective, SpyOnDirective] });
ScrollSpyModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: ScrollSpyModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: ScrollSpyModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [SpyTargetDirective, SpyOnDirective],
                    imports: [],
                    exports: [SpyTargetDirective, SpyOnDirective]
                }]
        }] });
