// ng-spy/esm2020/lib/spy-on.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./scroll-spy.service";
export class SpyOnDirective {
    constructor(el, renderer, spyService) {
        this.el = el;
        this.renderer = renderer;
        this.spyService = spyService;
        this.isActive = false;
    }
    ngOnInit() {
        this.spyService.activeSpyTarget.subscribe((targetName) => {
            if (!this.isActive && targetName === this.spyOn) {
                this.setActive();
            }
            else if (this.isActive && targetName !== this.spyOn) {
                this.setInActive();
            }
        });
    }
    get htmlElement() {
        return this.el.nativeElement;
    }
    setActive() {
        this.isActive = true;
        if (this.activeClass) {
            this.renderer.addClass(this.htmlElement, this.activeClass);
        }
    }
    setInActive() {
        this.isActive = false;
        if (this.activeClass) {
            this.renderer.removeClass(this.htmlElement, this.activeClass);
        }
    }
}
SpyOnDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: SpyOnDirective, deps: [{ token: i0.ElementRef }, { token: i0.Renderer2 }, { token: i1.ScrollSpyService }], target: i0.ɵɵFactoryTarget.Directive });
SpyOnDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.1.0", type: SpyOnDirective, selector: "[spyOn]", inputs: { activeClass: "activeClass", spyOn: "spyOn" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: SpyOnDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[spyOn]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i0.Renderer2 }, { type: i1.ScrollSpyService }]; }, propDecorators: { activeClass: [{
                type: Input
            }], spyOn: [{
                type: Input
            }] } });
