// ng-spy/esm2020/lib/spy-target.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./scroll-spy.service";
export class SpyTargetDirective {
    constructor(el, spyService, renderer) {
        this.el = el;
        this.spyService = spyService;
        this.renderer = renderer;
    }
    ngOnInit() {
        this.renderer.setAttribute(this.htmlElement, 'id', this.spyTarget);
        this.spyService.addTarget({ name: this.spyTarget, element: this.el });
    }
    get htmlElement() {
        return this.el.nativeElement;
    }
    ngOnDestroy() {
        this.spyService.removeTarget(this.spyTarget);
    }
}
SpyTargetDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: SpyTargetDirective, deps: [{ token: i0.ElementRef }, { token: i1.ScrollSpyService }, { token: i0.Renderer2 }], target: i0.ɵɵFactoryTarget.Directive });
SpyTargetDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.1.0", type: SpyTargetDirective, selector: "[spyTarget]", inputs: { spyTarget: "spyTarget" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: SpyTargetDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[spyTarget]'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.ScrollSpyService }, { type: i0.Renderer2 }]; }, propDecorators: { spyTarget: [{
                type: Input
            }] } });
