// ng-spy/esm2020/lib/window.service.mjs
import { isPlatformBrowser } from '@angular/common';
import { Inject, Injectable, PLATFORM_ID } from '@angular/core';
import { EMPTY, fromEvent } from 'rxjs';
import { auditTime } from 'rxjs/operators';
import { RESIZE_TIME_THRESHOLD } from './tokens/resize-threshold.token';
import { SCROLL_TIME_THRESHOLD } from './tokens/scroll-threshold.token';
import * as i0 from "@angular/core";
export class WindowService {
    constructor(platformId, resizeTime, scrollTime) {
        this.resizeTime = resizeTime;
        this.scrollTime = scrollTime;
        this.isBrowser = true;
        if (!isPlatformBrowser(platformId)) {
            this.isBrowser = false;
            this.scrollEvent$ = this.resizeEvent$ = EMPTY;
        }
        else {
            this.scrollEvent$ = fromEvent(window, 'scroll', { passive: true }).pipe(auditTime(this.scrollTime));
            this.resizeEvent$ = fromEvent(window, 'resize', { passive: true }).pipe(auditTime(this.resizeTime));
        }
    }
    getScrollEventForContainer(scrollContainer) {
        if (!this.isBrowser) {
            return EMPTY;
        }
        return fromEvent(scrollContainer.nativeElement, 'scroll', { passive: true }).pipe(auditTime(this.scrollTime));
    }
    get scrollEvent() {
        return this.scrollEvent$;
    }
    get resizeEvent() {
        return this.resizeEvent$;
    }
    get scrollTop() {
        if (!this.isBrowser) {
            return 0;
        }
        return Math.max(window.pageYOffset, document.documentElement.scrollTop, document.body.scrollTop);
    }
    get viewportHeight() {
        if (!this.isBrowser) {
            return 0;
        }
        return Math.max(document.documentElement.clientHeight, window.innerHeight || 0);
    }
    getElementHeight(el) {
        if (!this.isBrowser) {
            return 0;
        }
        return el.nativeElement.offsetHeight;
    }
    getElementOffsetTop(el) {
        if (!this.isBrowser) {
            return 0;
        }
        return el.nativeElement.offsetTop;
    }
    getElementScrollTop(el) {
        if (!this.isBrowser) {
            return 0;
        }
        return el.nativeElement.scrollTop;
    }
}
WindowService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: WindowService, deps: [{ token: PLATFORM_ID }, { token: RESIZE_TIME_THRESHOLD }, { token: SCROLL_TIME_THRESHOLD }], target: i0.ɵɵFactoryTarget.Injectable });
WindowService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: WindowService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.1.0", ngImport: i0, type: WindowService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root'
                }]
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [PLATFORM_ID]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [RESIZE_TIME_THRESHOLD]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [SCROLL_TIME_THRESHOLD]
                }] }]; } });
